README.md
# StreamPipe
**StreamPipe** is a simple Python library (a single file!) for running **multi-worker pipelines** on **infinite data streams**. The best part about StreamPipe is that the **order of data is preserved** throughout the pipeline. You can chain order-independent workers (think of image processing functions) and order-dependent workers (think of an object video tracker) on the same StreamPipe!

<p align="center">
  <img src="https://github.com/highvight/streampipe/assets/25797927/95e5a820-eb1a-43f7-ad67-8bd8ff94fb12" width="400" align="center">
</p>



## Installation
```bash
pip3 install streampipe
```

Install from source:
```bash
git clone https://github.com/highvight/streampipe
cd streampipe
pip3 install .
pip3 install .[testing]  # for testing and benchmarks
```

## Usage
Here is a starter. Let's check if you can speed up a reshaping and standardization pipeline for images.

```python
import timeit
import numpy as np
from streampipe import StreamPipe

IMG = np.random.rand(1000, 1000, 3) * 255

def _reshape(img):
    return img[::2, ::2]

def _standardize(img):
    mean = np.mean(img, axis=(0, 1), keepdims=True)
    std = np.std(img, axis=(0, 1), keepdims=True)
    return (img - mean) / std

pipe = StreamPipe()
pipe.add(_reshape, n_workers=4, maxsize=4)
pipe.add(_standardize, n_workers=4, maxsize=4)

def measure_pipe():
    # Starts the workers
    with pipe:
        for _ in range(100):
            pipe.put(IMG)
    # All workers stopped, all items done

def measure_normal():
    for _ in range(100):
        x = _reshape(IMG)
        x = _standardize(x)

print(timeit.timeit(measure_pipe, number=1))  # 1.214164252000046
print(timeit.timeit(measure_normal, number=1))  # 3.083142001007218
```
    
## License

[MIT](https://choosealicense.com/licenses/mit/)


CHANGELOG.rst
=========
Changelog
=========

Version 0.1
===========

- Init release


tests/test_seqpipe.py
import threading
from time import sleep
from timeit import default_timer

import pytest

from streampipe import StreamPipe


__author__ = "Adrian Krueger"
__copyright__ = "Adrian Krueger"
__license__ = "MIT"


def _identity(x):
    return x


def _square(x):
    return x**2


def _root(x):
    return x**0.5


def _append(x, l):
    l.append(x)
    return x


def _error_on_2(x):
    if x == 2:
        raise Exception()
    return x


def test_sample_run():
    target = [float(i) for i in range(100)]
    test = []

    streampipe = StreamPipe()
    streampipe.add(_square, n_workers=5, maxsize=3)
    streampipe.add(_identity, n_workers=5, maxsize=3)
    streampipe.add(_root, n_workers=5, maxsize=3)
    streampipe.add(_append, n_workers=1, maxsize=2, args=(test,))

    streampipe.start()
    for i in range(100):
        streampipe.put(i)
    streampipe.join_queues()
    streampipe.stop()
    streampipe.join_workers()

    assert target == test


def test_order():
    target = [i for i in range(100)]
    test = []

    streampipe = StreamPipe()
    streampipe.add(_identity, n_workers=10, maxsize=3)
    streampipe.add(_identity, n_workers=10, maxsize=3)
    streampipe.add(_identity, n_workers=10, maxsize=3)
    streampipe.add(_identity, n_workers=10, maxsize=3)
    streampipe.add(_identity, n_workers=10, maxsize=3)
    streampipe.add(_identity, n_workers=10, maxsize=3)
    streampipe.add(_append, n_workers=1, maxsize=2, args=(test,))

    with streampipe:
        for i in range(100):
            streampipe.put(i)

    assert test == target


@pytest.mark.filterwarnings("ignore::pytest.PytestUnhandledThreadExceptionWarning")
def test_target_error():
    streampipe = StreamPipe()
    streampipe.add(_identity, n_workers=2, maxsize=2)
    streampipe.add(_error_on_2, n_workers=2, maxsize=2)

    with streampipe:
        streampipe.put(1)  # ok
        streampipe.put(2)  # worker fails

    assert len(list(threading.enumerate())) == 1
    streampipe.join_workers()  # should be finished
    with pytest.raises(Exception):
        streampipe.start()


def test_delete():
    streampipe = StreamPipe()
    streampipe.add(_identity, n_workers=2, maxsize=2)
    streampipe.start()

    start = default_timer()
    while len(list(threading.enumerate())) != 3:
        sleep(0.01)
        assert default_timer() - start < 1

    streampipe.stop()

    start = default_timer()
    while len(list(threading.enumerate())) != 1:
        sleep(0.01)
        assert default_timer() - start < 1


benchmark/test_benchmark_seqpipe.py
import pytest

import numpy as np
from numpy.random import rand

from streampipe import StreamPipe


__author__ = "Adrian Krueger"
__copyright__ = "Adrian Krueger"
__license__ = "MIT"


ROUNDS = 5
N_IMAGES = 100


def _reshape(img):
    return img[::2, ::2]


def _standardize(img):
    mean = np.mean(img, axis=(0, 1), keepdims=True)
    std = np.std(img, axis=(0, 1), keepdims=True)
    return (img - mean) / std


def _test_baseline(img: np.array):
    for _ in range(N_IMAGES):
        x = _reshape(img)
        x = _standardize(x)


def _test_pipe(pipe: StreamPipe, img: np.array):
    for _ in range(N_IMAGES):
        pipe.put(img)
    pipe.join_queues()
    pipe.stop()


@pytest.mark.parametrize("n_workers", [""])
@pytest.mark.parametrize("maxsize", [""])
def test_baseline(benchmark, n_workers, maxsize):
    img = rand(1000, 1000, 3) * 255
    benchmark.pedantic(_test_baseline, args=(img,), rounds=ROUNDS)


@pytest.mark.parametrize("n_workers", [1, 4, 8])
@pytest.mark.parametrize("maxsize", [1, 4, 8])
def test_pipe_multithread(benchmark, maxsize, n_workers):
    def setup():
        img = rand(1000, 1000, 3) * 255
        pipe = StreamPipe()
        pipe.add(_reshape, n_workers=n_workers, maxsize=maxsize)
        pipe.add(_standardize, n_workers=n_workers, maxsize=maxsize)
        pipe.start()
        return (pipe, img), {}

    benchmark.pedantic(_test_pipe, setup=setup, rounds=ROUNDS)


# Verz slow multiprocessing performance

# @pytest.mark.parametrize("n_workers", [1, 4, 8])
# @pytest.mark.parametrize("maxsize", [1,4,8])
# def test_pipe_multiprocess(benchmark, maxsize, n_workers):
#     def setup():
#         img = rand(1000, 1000, 3) * 255
#         pipe = StreamPipe()
#         pipe.add(_reshape, n_workers=n_workers, maxsize=maxsize, multiprocessing=True)
#         pipe.add(_standardize, n_workers=n_workers, maxsize=maxsize, multiprocessing=True)
#         pipe.start()
#         return (pipe, img), {}

#     benchmark.pedantic(_test_pipe, setup=setup, rounds=ROUNDS)


src/streampipe/streampipe.py
import multiprocessing as mp
from multiprocessing import Process
from multiprocessing import Event as pEvent
from queue import Empty, Full
from time import sleep
from threading import Thread
from threading import Event as tEvent
from typing import Any, Callable, Iterable, Mapping
import warnings

from seqqueue import SeqQueue

__author__ = "Adrian Krueger"
__copyright__ = "Adrian Krueger"
__license__ = "MIT"


class ThreadWorker(Thread):
    def __init__(
        self,
        target: Callable[..., object],
        in_q: SeqQueue,
        out_q: SeqQueue | None,
        stop_event: tEvent,
        error_event: tEvent,
        group: None = None,
        name: str | None = None,
        args: Iterable[Any] = (),
        kwargs: Mapping[str, Any] | None = None,
        *,
        daemon: bool | None = None,
    ) -> None:
        super().__init__(group, target, name, args, kwargs, daemon=daemon)
        self.in_q = in_q
        self.out_q = out_q
        self.stop_event = stop_event
        self.error_event = error_event

    def run(self):
        try:
            while True:
                # Get item from in_q
                while not self.stop_event.is_set():
                    try:
                        index, item = self.in_q.get(block=True, timeout=0.1)
                        break
                    except Empty:
                        continue
                else:
                    return

                # Apply target function
                index, item = index, self._target(item, *self._args, **self._kwargs)

                # Put item to out_q
                if self.out_q is not None:
                    while not self.stop_event.is_set():
                        try:
                            self.out_q.put((index, item), block=True, timeout=0.1)
                            break
                        except Full:
                            continue
                    else:
                        return

                self.in_q.task_done()

        except Exception as e:
            try:
                self.error_event.set()
                # Notify other to stop
                self.stop_event.set()
            finally:
                raise (e)
        finally:
            # Avoid a refcycle if the thread is running a function with
            # an argument that has a member that points to the thread.
            del self._target, self._args, self._kwargs


class ProcessWorker(Process):
    def __init__(
        self,
        target: Callable[..., object],
        in_q: SeqQueue,
        out_q: SeqQueue | None,
        stop_event: pEvent,
        error_event: pEvent,
        group: None = None,
        name: str | None = None,
        args: Iterable[Any] = (),
        kwargs: Mapping[str, Any] = {},
        *,
        daemon: bool | None = None,
    ) -> None:
        super().__init__(group, target, name, args, kwargs, daemon=daemon)
        self.in_q = in_q
        self.out_q = out_q
        self.stop_event = stop_event
        self.error_event = error_event

    def run(self):
        try:
            while True:
                # Get item from in_q
                while not self.stop_event.is_set():
                    try:
                        index, item = self.in_q.get(block=True, timeout=0.1)
                        break
                    except Empty:
                        continue
                else:
                    return

                # Apply target function
                index, item = index, self._target(item, *self._args, **self._kwargs)

                # Put item to out_q
                if self.out_q is not None:
                    while not self.stop_event.is_set():
                        try:
                            self.out_q.put((index, item), block=True, timeout=0.1)
                            break
                        except Full:
                            continue
                    else:
                        return

                self.in_q.task_done()

        except Exception as e:
            try:
                self.error_event.set()
                # Notify others to stop
                self.stop_event.set()
            finally:
                raise (e)


class StreamPipe:
    """StreamPipe enables sequential, multi-worker manipulation of incoming data.

    >>> # Print 0,...,99
    >>> pipe = StreamPipe()
    >>> pipe.add(lambda x: x**0.5, n_workers=5, maxsize = 5)
    >>> pipe.add(lambda x: print(x), n_workers=1, maxsize = 5)
    >>> with pipe:
    >>>     for i in range(100):
    >>>         pipe.put(i**2)

    """

    def __init__(self) -> None:
        self.index = 0
        self._worker_groups: Iterable[Iterable[ThreadWorker | ProcessWorker]] = []
        self._stop_event = tEvent()
        self._error_event = tEvent()
        self._manager = None

    def put(self, item: Any, block: bool = True, timeout: float | None = None) -> None:
        """Put an item on the StreamPipe.

        Works just like queue.put(). Empty and Full exceptions/blocks are raised with
        respect to the 'maxsize' of the first added worker group.

        Args:
            item (Any): The item to put on the StreamPipe.
            block (bool, optional): Block until a slot becomes available at most 'timeout'
                seconds. Defaults to True.
            timeout (float | None, optional): Maximum block timeout in seconds.
                Defaults to None.

        """
        if not self._worker_groups:
            # No worker has been added, do nothing
            return
        self._worker_groups[0][0].in_q.put((self.index, item), block, timeout)
        self.index += 1

    def add(
        self,
        target: Callable,
        n_workers: int,
        maxsize: int = 0,
        multiprocessing: bool = False,
        args: Iterable[Any] = (),
        kwargs: Mapping[str, Any] | None = None,
    ) -> None:
        """Add a worker group to the StreamPipe.

        A worker group consists of 'n_workers' concurrent workers running 'target'
        on the next incoming item. The item must be the first argument of 'target'.
        If subsequent worker groups are added, 'target' must return the next item.

        >>> # Add a worker group of 5 returning the square of x.
        >>> pipe.add(lambda x: x**2, n_workers=5, maxsize = 3)
        >>> # For multiprocessing workers
        >>> pipe.add(lambda x: print(x), n_workers=1, maxsize = 3, multiprocessing+True)

        Args:
            target (Callable): The target function of the workers.
            n_workers (int): The number of concurrent workers executing 'target' on
                incoming data.
            maxsize (int, optional): The maxsize of the incoming data queue.
                Defaults to 0.
            multiprocessing (bool, optional): If True, workers run in seperate
                processes (rather than threads). Defaults to False.
            args (Iterable[Any], optional): Addtional Args passed to 'target'.
                Defaults to ().
            kwargs (Mapping[str, Any] | None, optional): Kwargs passed to 'target'.
                Defaults to None.
        """

        if n_workers < 1:
            return

        # Create worker group
        if multiprocessing:
            self._init_mp()
            in_q = self._manager.SeqQueue(maxsize=maxsize, start_index=self.index)
            worker_group = [
                ProcessWorker(
                    target=target,
                    in_q=in_q,
                    out_q=None,
                    stop_event=self._stop_event,
                    error_event=self._error_event,
                    args=args,
                    kwargs=kwargs if kwargs is not None else {},
                    name=f"{target.__name__}_worker{len(self._worker_groups)}_{i}",
                    daemon=True,
                )
                for i in range(n_workers)
            ]
        else:
            if len(self._worker_groups) > 0 and not isinstance(
                self._worker_groups[-1][0].in_q, SeqQueue
            ):
                # The preceding worker group has multiprocessing
                in_q = self._manager.SeqQueue(maxsize=maxsize, start_index=self.index)
            else:
                in_q = SeqQueue(maxsize=maxsize, start_index=self.index)
            worker_group = [
                ThreadWorker(
                    target=target,
                    in_q=in_q,
                    out_q=None,
                    stop_event=self._stop_event,
                    error_event=self._error_event,
                    args=args,
                    kwargs=kwargs,
                    name=f"{target.__name__}_worker{len(self._worker_groups)}_{i}",
                    daemon=True,
                )
                for i in range(n_workers)
            ]

        # Attach to previous worker group (if any)
        if len(self._worker_groups) > 0:
            for worker in self._worker_groups[-1]:
                worker.out_q = in_q

        # Add to worker groups
        self._worker_groups.append(worker_group)

    def start(self):
        """Start all workers.

        Raises:
            ValueError: If StreamPipe has been stopped.
        """
        if self._stop_event.is_set():
            raise RuntimeError("StreamPipe cannot be restarted")

        for worker_group in self._worker_groups:
            for worker in worker_group:
                worker.start()

    def stop(self):
        """Stop all workers."""
        self._stop_event.set()

    def join_queues(self):
        """Block until all current tasks on the StreamPipe are finished."""
        for worker_group in self._worker_groups:
            worker_group[0].in_q.join()

    def join_workers(self):
        """Block until all workers are stopped."""
        for worker_group in self._worker_groups:
            for worker in worker_group:
                worker.join()

    def unfinished_tasks(self) -> int:
        """Return the total number of unfinished tasks on the StreamPipe.

        Returns:
            int: Total number of unfinished tasks
        """
        unfinished_tasks = 0
        for worker_group in self._worker_groups:
            unfinished_tasks += worker_group[0].in_q.unfinished_tasks
        return unfinished_tasks

    def _init_mp(self):
        if self._manager is None:
            self._manager = mp.Manager()
            self._stop_event = self._manager.Event()
            self._error_event = self._manager.Event()
            for worker_group in self._worker_groups:
                for worker in worker_group:
                    worker.stop_event = self._stop_event
                    worker.error_event = self._error_event

    def __enter__(self):
        self.start()
        return self

    def __exit__(self, exc_type, exc, exc_tb):
        # Wait until all unfinished tasks are completed
        while self.unfinished_tasks() > 0:
            if self._stop_event.is_set():
                self.join_workers()
                if self._error_event.is_set():
                    warnings.warn(
                        f"StreamPipe stopped due to a worker error. A total of {self.unfinished_tasks()} tasks could not be completed",
                        RuntimeWarning,
                    )
                else:
                    warnings.warn(
                        f"StreamPipe has been stopped, but a total of {self.unfinished_tasks()} tasks could not be completed",
                        RuntimeWarning,
                    )
                return
            sleep(0.01)
        # Stop StreamPipe and wait until all workers are shutdown.
        self.stop()
        self.join_workers()
